### PR TITLE
+Added VEL_UNDERFLOW and STRESS_UNDERFLOW

### PR DESCRIPTION
**Pull request by Hallberg-NOAA**
Added the new runtime parameters VEL_UNDERFLOW and STRESS_UNDERFLOW to set
minimal values of the velocity and ice-pack lateral stresses that lead to an
underflow of the velocities and stresses to zero.  This new capability can be
needed to give identical results when dimensional consistency tests are carried
out.  Also added the runtime parameter B_DYNAMICS_MIN_MASS to replace a hard
coded value for the B-grid dynamics code.  By default, all answers are bitwise
identical, but there are 2 or 3 new runtime parameters in the SIS_parameter_doc
files.

### CODE REVIEW

**Comment by marshallward** [comment]
It looks like some of these if-blocks could be moved outside of the loop.  Am I right?
If so, then it would be better for performance.

**Comment by Hallberg-NOAA** [comment]
The if-blocks are being applied individually at every point, and hence these tests have to be inside of the loops.

  It would be possible to add if-tests for whether the underflow values are greater than zero, and then if this is true add an extra set of specific loops within which to do the tests on each point, but if the underflow values are positive I don't see how this would improve performance.  Answers would be unchanged if it were altered this way, and I would expect a sufficiently clever compiler (is that an oxymoron?) might do this rearrangement itself.

**Comment by marshallward** [comment]
Yes, that is what I am thinking.  Split it into two do-loops, and wrap the second one in a global if-block.

At the very least, the if-block operations interrupt the arithmetic pipeline, so I do believe that it will impede performance.

I don't necessarily think a compiler would opt to split this loop.  Instruction scheduling is generally very difficult and I would not personally recommend that we count on it.  I suppose we could experiment.  But I would be more comfortable moving it outside of the block.

**Comment by Hallberg-NOAA** [comment]
The updated code goes about as far as I think is safe in moving these underflow tests out of the loops.  With the B-grid version of the rheology solver, these velocities are being reused extensively after being set, so I don't know that it is safe to reorder the calculations there, but we don't make much use of the B-grid version anyway.

File: src/SIS_dyn_cgrid.F90 [hunk 1]
```diff
@@ -2083,10 +2124,13 @@ subroutine SIS_C_dyn_read_alt_restarts(CS, G, US, Ice_restart, restart_dir)
                      (US%kg_m3_to_R_restart * US%m_to_Z_restart * US%m_to_L_restart**2)
     do J=G%jsc-1,G%jec ; do I=G%isc-1,G%iec
       CS%str_s(I,J) = stress_rescale * CS%str_s(I,J)
+      if (abs(CS%str_s(I,J)) < CS%str_underflow) CS%str_s(I,J) = 0.0
```
[the remark this answers is not in the document]
Comment (Hallberg-NOAA) [reply src/SIS_dyn_cgrid.F90:2127]: I omitted that one because it only occurs at initialization from an older-format restart file that will no longer be written for versions of the code subsequent to https://github.com/NOAA-GFDL/SIS2/pull/175.

**Review by marshallward — APPROVED** [review]

**Comment by marshallward** [comment]
Gaea regression: https://gitlab.gfdl.noaa.gov/ogrp/SIS2/-/pipelines :heavy_check_mark: :yellow_circle:
